libbeat/publisher/queue/diskqueue/segments.go

```diff
@@ -128,42 +147,59 @@ func (s bySegmentID) Less(i, j int) bool { return s[i].id < s[j].id }
 
 // Scan the given path for segment files, and return them in a list
 // ordered by segment id.
-func scanExistingSegments(path string) ([]*queueSegment, error) {
-	files, err := ioutil.ReadDir(path)
+func scanExistingSegments(logger *logp.Logger, pathStr string) ([]*queueSegment, error) {
+	files, err := ioutil.ReadDir(pathStr)
 	if err != nil {
-		return nil, fmt.Errorf("Couldn't read queue directory '%s': %w", path, err)
+		return nil, fmt.Errorf("couldn't read queue directory '%s': %w", pathStr, err)
 	}
 
 	segments := []*queueSegment{}
 	for _, file := range files {
-		if file.Size() <= segmentHeaderSize {
-			// Ignore segments that don't have at least some data beyond the
-			// header (this will always be true of segments we write unless there
-			// is an error).
-			continue
-		}
 		components := strings.Split(file.Name(), ".")
 		if len(components) == 2 && strings.ToLower(components[1]) == "seg" {
 			// Parse the id as base-10 64-bit unsigned int. We ignore file names that
 			// don't match the "[uint64].seg" pattern.
 			if id, err := strconv.ParseUint(components[0], 10, 64); err == nil {
-				segments = append(segments,
-					&queueSegment{
-						id:        segmentID(id),
-						endOffset: segmentOffset(file.Size() - segmentHeaderSize),
-					})
+				fullPath := path.Join(pathStr, file.Name())
+				header, err := readSegmentHeaderWithFrameCount(fullPath)
+				if header == nil {
+					logger.Errorf("couldn't load segment file '%v': %v", fullPath, err)
```

Comment:
I'd like to encourage us to use structured logging more. e.g. the loop could introduce its own logger like `logger := logger.With("segment", segmentID(id))`. Would we need the full path if we have had the segment ID?

Reply:
The full path is theoretically redundant since every segment is in the same directory, and the directory depends on the user configuration, but I thought it would be nice for the error to be explicit since I wouldn't expect whoever sees the message to know where the queue path is or how to find it.